### PR TITLE
Replay logging

### PR DESCRIPTION
**Pull request by cnr**
This PR introduces "replay logging", i.e., the capability to record and replay effects that are run.

## Effect recording - `Control.Effect.Record`/`Control.Effect.Record.TH`
These additions allow us to record the effect invocations of arbitrary effects into a JSON structure. Any first-order effect can be recorded, i.e., any effect that doesn't use the `m` parameter in the effect constructors.

For example, this is fine:
```hs
data TTY m a where
  ReadTTY :: TTY m String
  WriteTTY :: String -> TTY m ()
```

..but this is not:
```hs
data Alt m a where
  Alt :: m a -> m a -> Alt m a
```

Effects we require a `Recordable` instance, but in general, this can be freely derived with TemplateHaskell via `deriveRecordable` from `Control.Effect.Record.TH`:
```hs
$(deriveRecordable ''TTY)
```

With a `Recordable` instance available, recording an effect is as easy as, e.g., calling `runRecord @TTY`, which produces a `Journal` (which has a `ToJSON` instance)

## Effect replaying - `Control.Effect.Replay`/`Control.Effect.Replay.TH`
These additions allow us to use the `Journal` we recorded to replay effects.

In addition to a `Recordable` instance, effects we're replaying require a `Replayable` instance, which in general can be freely derived with TemplateHaskell via `deriveReplayable` from `Control.Effect.Replay.TH`

With a `Replayable` instance available, replaying an effect can be done through, e.g., `runReplay @TTY journal`

### CODE REVIEW

File: src/Effect/ReadFS.hs [hunk 1]
```diff
@@ -187,6 +225,9 @@ instance Has (Lift IO) sig m => Algebra (ReadFS :+: sig) (ReadFSIOC m) where
       -- NB: these never throw
       L (DoesFileExist file) -> (<$ ctx) <$> sendIO (PIO.doesFileExist file)
       L (DoesDirExist dir) -> (<$ ctx) <$> sendIO (PIO.doesDirExist dir)
+      L (ListDir dir) -> do
+        res <- catchingIO (PIO.listDir dir) (ListDirError (toFilePath dir))
+        pure (res <$ ctx)
```
Comment (scruffystuffs) [inline src/Effect/ReadFS.hs:230]: Should we move this above the note about never throwing?

**Review by scruffystuffs — CHANGES_REQUESTED** [review]
Everything looks good.  Only changes needed are in `analyzeMain`.  If possible, I'd like to see logging and diagnostics move out of the `if->then->else` block in that function.  Also, you may want to try rebasing onto the current master.

File: src/Strategy/NuGet/Nuspec.hs [hunk 2]
```diff
@@ -42,14 +41,14 @@ data NuspecProject = NuspecProject
...
+mkProject :: (Has ReadFS sig n, Has Diagnostics sig n) => NuspecProject -> DiscoveredProject n
 mkProject project =
   DiscoveredProject
     { projectType = "nuspec",
       projectBuildTargets = mempty,
-      projectDependencyGraph = const . runReadFSIO $ getDeps project,
+      projectDependencyGraph = const $ getDeps project,
       projectPath = parent $ nuspecFile project,
-      projectLicenses = runReadFSIO $ analyzeLicenses (nuspecFile project)
+      projectLicenses = analyzeLicenses (nuspecFile project)
```
[the remark this answers is not in the document]
Comment (cnr) [reply src/Strategy/NuGet/Nuspec.hs:51]: This comes from `pathfinder`, which we're looking to phase out. Eventually we'd like to eliminate `analyzeLicenses` entirely

**Review by scruffystuffs — APPROVED** [review]
This whole PR is awesome!  I'm stoked to have this.

giphub

**Pull request merged**